src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts

```diff
@@ -0,0 +1,172 @@
+		await this.updateOutlineModel(this.cts.token);
+		this.onStickyScrollChangeEmitter.fire();
+	}
+
+	private async updateOutlineModel(token: CancellationToken) {
+		if (this.editor.hasModel()) {
+			const model = this.editor.getModel();
+			const modelVersionId = model.getVersionId();
+			const outlineModel = await OutlineModel.create(this.languageFeaturesService.documentSymbolProvider, model, token) as OutlineModel;
+			if (token.isCancellationRequested) {
+				return;
+			}
+			this.outlineModel = StickyOutlineElement.fromOutlineModel(outlineModel);
+			this.modelVersionId = modelVersionId;
+		}
+	}
+
+	public getCandidateStickyLinesIntersectingFromOutline(range: StickyRange, outlineModel: StickyOutlineElement, result: StickyLineCandidate[], depth: number, lastStartLineNumber: number): void {
+		let lastLine = lastStartLineNumber;
+		for (const child of outlineModel.children) {
+			if (child.range) {
+				const childStartLine = child.range.startLineNumber;
+				const childEndLine = child.range.endLineNumber;
+				if (range.startLineNumber <= childEndLine + 1 && childStartLine - 1 <= range.endLineNumber && childStartLine !== lastLine) {
+					lastLine = childStartLine;
+					result.push(new StickyLineCandidate(childStartLine, childEndLine - 1, depth + 1));
+					this.getCandidateStickyLinesIntersectingFromOutline(range, child, result, depth + 1, childStartLine);
+				}
+			} else {
+				this.getCandidateStickyLinesIntersectingFromOutline(range, child, result, depth, lastStartLineNumber);
+			}
+		}
+	}
+
+	public getCandidateStickyLinesIntersecting(range: StickyRange): StickyLineCandidate[] {
+		let stickyLineCandidates: StickyLineCandidate[] = [];
+		if (range.startLineNumber && range.endLineNumber) {
+			this.getCandidateStickyLinesIntersectingFromOutline(range, this.outlineModel as StickyOutlineElement, stickyLineCandidates, 0, -1);
+		}
+		const hiddenRanges: Range[] | undefined = this.editor._getViewModel()?.getHiddenAreas();
+		if (hiddenRanges) {
+			for (const hiddenRange of hiddenRanges) {
+				stickyLineCandidates = stickyLineCandidates.filter(stickyLine => !(stickyLine.startLineNumber >= hiddenRange.startLineNumber && stickyLine.endLineNumber <= hiddenRange.endLineNumber + 1));
+			}
+		}
+		return stickyLineCandidates;
+	}
+
+	override dispose(): void {
+		super.dispose();
+		this.sessionStore.dispose();
+	}
+}
+
+class StickyOutlineElement {
+	public static fromOutlineModel(outlineModel: OutlineModel | OutlineElement | OutlineGroup): StickyOutlineElement {
+		const children = [...outlineModel.children].map(entry =>
+			StickyOutlineElement.fromOutlineModel(entry[1])
+		);
+		children.sort((child1, child2) => {
+			if (!child1.range || !child2.range) {
+				return 1;
+			} else if (child1.range.startLineNumber !== child2.range.startLineNumber) {
+				return child1.range.startLineNumber - child2.range.startLineNumber;
+			} else {
+				return child2.range.endLineNumber - child1.range.endLineNumber;
+			}
+		});
+		let range;
```

Comment:
This will have type `any`. You could use `let range: StickyRange | undefined` to avoid that.